perf: de-prioritize rendering non-visible tiles

sgothel opened the pull request:
* Resolves: #9989
* Target version: master 

### Summary
Render combined-tiles, timeout limited, invisible last
    
Requested combined-tiles are send until timeout,
at least one item, visible items first.
This approach shall increase responsiveness.

`KitSocketPoll::kitPoll` loops through max-clipped 100ms `poll` slices,
allowing new incoming client requests.
`TileCombined` request are sent as much as remaining time allows,
at least on item while favoring visible items.

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

sgothel commented:
The TileCombined's dimension, i.e. its _width and _height, is not fixed in the streaming TileCombined  ctor (line ~404).
Only the 1st TileDesc dimenstion is being stored.
The combined dimension seems to be used at doRender only?
If fixed, i.e. AABBox style grown (-> Rectangle) we simply could use it to determine visibility .. instead of looping through all recreated TileDesc instances.

sgothel commented:
I need clarification regarding the IDLE criteria, which shall allow sending the invisible tiles.
(timeout  = to)
This patch uses the to_left < time_consumed in drainedQueue, which works if initial to is reasonable reflecting time left.
Testing showed that a to > 0 is only given for IDLE periods?

sgothel commented:
Force push resolved merge conflict with latest master tip as detected with the bots. I must have pushed it this morning in-between.

mmeeks commented on common/Rectangle.hpp:
```diff
@@ -63,39 +85,49 @@ struct Rectangle
-        intersection._x2 = std::min(_x2, rOther._x2);
-        intersection._y2 = std::min(_y2, rOther._y2);
-        return intersection.isValid();
+        const int x1 = std::max(_x1, rOther._x1);
```
If there are no unit tests for this code; why change it ? If we are changing it in a not totally obvious way - there should be unit tests.

sgothel replied:
WhiteBoxTests.cpp does validate Rectangle intersects after consolidation via TileDesc::rectanglesIntersect

mmeeks commented on kit/ChildSession.cpp:
```diff
@@ -1223,8 +1223,7 @@ bool ChildSession::clientVisibleArea(const StringVector& tokens)
 {
-    return (tile.getTilePosX() >= _clientVisibleArea.getLeft() && tile.getTilePosX() <= _clientVisibleArea.getRight() &&
-        tile.getTilePosY() >= _clientVisibleArea.getTop() && tile.getTilePosY() <= _clientVisibleArea.getBottom());
+    return tile.intersects( _clientVisibleArea );
```
Looks prettier - does that exchange four comparisons for four branches with std::min and two comparisons ? or did I miss something ? =)

mmeeks commented on kit/Kit.cpp:
```diff
@@ -2296,7 +2296,7 @@ void Document::drainCallbacks()
 }
 
-void Document::drainQueue()
+void Document::drainQueue(std::chrono::steady_clock::time_point now, int timeoutMicroS)
```
Like the passing of now =)

mmeeks commented on kit/Kit.cpp:
```diff
@@ -2365,11 +2365,40 @@ void Document::drainQueue()
+                                           std::chrono::steady_clock::now() - now).count();
+            timeoutMicroS -= consumedMicroS;
+
+            if( timeoutMicroS >= consumedMicroS )
```
This is curious to me. If we have a huge batch of tiles that will take 100's of milliseconds to render - some of them visible, some of them irrelevant - we will now go render all of them in if it didn't take very long to process the preceeding input events (?)

But we will not re-spin the poll to improve responsiveness or interactivity in this case - right ?

mmeeks commented on kit/Kit.cpp:
```diff
@@ -2365,11 +2365,40 @@ void Document::drainQueue()
+                {
+                    if( isTileRequestInsideVisibleArea(req) )
+                    {
+                        renderTiles(req);
```
Here we seem to render tiles individually one by one - which can create a massive performance problem - vs. rendering groups of them; why do we do that ?

mmeeks commented on wsd/ClientSession.cpp:
```diff
@@ -3012,9 +3011,8 @@ bool ClientSession::isTileInsideVisibleArea(const TileDesc& tile) const
-        if (tile.getTilePosX() >= paneRect.getLeft() && tile.getTilePosX() <= paneRect.getRight() &&
-            tile.getTilePosY() >= paneRect.getTop() && tile.getTilePosY() <= paneRect.getBottom())
+        const Util::Rectangle paneRect = getNormalizedVisiblePaneArea(panes[i]);
+        if( tile.intersects( paneRect ) )
```
easier to read I suppose; I wonder why we never did that - I suspect this method is called a lot; what does the history say ?

mmeeks reviewed (changes requested):
Thanks, some useful cleanups here perhaps - though the algorithm used doesn't seem to match what was asked for. Let me paste that again:

a) only render & send those that are inside a visible area and
b) re-spin the main-loop  and
c) only render non-visible tiles when there is nothing else to do =)

Was there a problem with detecting nothing else to do in c) - I would expect that is reasonably easy to detect - when we found nothing to read from our remote socket, and had a non-zero timeout to poll for.

"an attempt to reduce bandwidth to client" - this is primarily a way of reducing latency, by not spending time rendering not-yet-seen tiles - and prioritizing visible tiles over non-visible ones. Nothing to do with bandwidth in particular.

mmeeks commented:
It would be good to cleanup the ordering functionality at the same time though eg.

// FIXME: it's not that clear what good this does for us ...
// we process all previews in the same batch of rendering
void KitQueue::deprioritizePreviews()

Around thumbnailing slides - while ensuring that whatever algorithm we use doesn't create starvation problems.

sgothel commented on common/Rectangle.hpp:
```diff
@@ -63,39 +85,49 @@ struct Rectangle
     }
 
-    void setLeft(int x1) { _x1 = x1; }
+    constexpr void setLeft(int x1) /*noexcept*/ { _x1 = x1; }
```
I obeyed the `noexcept` rule, just wanted to annotate for `constexpr`.
Will drop this annotation.

Its is correct, that the compiler very highly likely will optimize all calls to become inline.
However, `constexpr` explicitly does
- enforced inline (`inline` itself is useless)
- enforced check/removal of UB (undefined behavior)
- allow compile time if possible (depending on data source, ...)
  - would require `noexcept` - hence the annotation (to be dropped).

Boils down to the difference of `can` and `must` + perhaps UB.

If you insist of dropping the `must inline` property of `constexpr` in general, I will ofc.

sgothel commented:
> It would be good to cleanup the ordering functionality at the same time though eg.
> 
> // FIXME: it's not that clear what good this does for us ... // we process all previews in the same batch of rendering void KitQueue::deprioritizePreviews()

Looking at it, it only pushes _leading_ preview-tiles to the end of the queue
and you mentioned in the comment that a tile-req batch either only consist out of 
such preview-tiles or none shall be included(?).

Our IDLE/invisible mechanism act above the produced 'popWholeTileQueue'
on the produced combined tiles, i.e. included this preview reordering.
Not sure if this is the discussion here.

> 
> Around thumbnailing slides - while ensuring that whatever algorithm we use doesn't create starvation problems.

Since out IDLE/invisible mechanism includes the preview-tile order, see above,
this should be well included in the solution.

sgothel commented:
Offered the optional Tile/TileCombined intersection test patch via
https://github.com/CollaboraOnline/online/pull/10197

sgothel commented:
COOL 24.04 Cypress (desktop) — FAILURE:
Passed locally: make check-desktop spec=writer/invalidations_spec.js

caolanm commented on kit/Kit.cpp:
```diff
@@ -2806,20 +2829,32 @@ int KitSocketPoll::kitPoll(int timeoutMicroS)
 
-            timeoutMicroS =
-                std::chrono::duration_cast<std::chrono::microseconds>(_pollEnd - now).count();
             ++eventsSignalled;
```
This ++eventSignalled line wasn't incremented in the before-change case of a poll returning 0, but it is in the after-case (similarly timeoutMicroS is set to 0) and eventSignalled and timeoutMicroS are used after the loop, so that would be a change. Do we want that?

Would there be any real difference in intent, on, say, just passing false to the first drainQueue inside the loop and passing an isIDLE to the trailing drainQueue outside the loop, where isIDLE is set when poll is passed a non-zero timeout and returns 0 ?

sgothel commented:
rebased to master a330f9e75717de96c72794d0ac672f75376c9711 (no changes)

sgothel commented:
re CodeQL see https://github.com/CollaboraOnline/online/pull/9916#issuecomment-2415664161

mmeeks commented on kit/Kit.cpp:
```diff
@@ -2300,7 +2300,7 @@ void Document::drainCallbacks()
 }
 
-void Document::drainQueue()
+void Document::drainQueue(bool isIDLE)
```
Why is IDLE capitalized. Parameter makes some sense: 'isIdle' is reasonable.

mmeeks commented on kit/Kit.cpp:
```diff
@@ -2822,15 +2845,24 @@ int KitSocketPoll::kitPoll(int timeoutMicroS)
                 realTimeout = 0;
 
-            if (poll(std::chrono::microseconds(realTimeout)) <= 0)
+            const int rc = poll(std::chrono::microseconds(realTimeout));
```
Not convinced by this. What 'idle' are we detecting here. We can tell when the core main-loop is idle from a non-zero timeout; that's fine - how do we detect our own ie. coolforkit idle-ness ? presumably with a very short poll timeout - is that what we do ? confused here. Looks like we conclude that if we waited a long time and timed-out, we are now idle - and now we can do the things we wanted to do but were delayed; but why delay them ? 

Confused.

mmeeks reviewed (changes requested):
Should prolly have a call here.

sgothel commented:
PR #10183 as well as PR #10512 were rebased (forced pushed) twice
- first rebase
  - fixed to the loIdle determination
  - Document::drainQueue returns DocDrainStats for logging details, 
    also reducing code duplication in KitSocketPoll::kitPoll
- second rebase
  - rebase to master

caolanm commented on kit/Kit.cpp:
```diff
@@ -2441,17 +2443,39 @@ void Document::drainQueue()
+            !isBackgroundSaveProcess() && tilesQueued > 0)
         {
+            // Render all tiles if idle.
+            // Otherwise render only the first visible TileCombined, but at least one TileCombined (if non are visible).
```
trivial nit: non->none

caolanm reviewed (commented):
I guess the obvious question then is if https://github.com/CollaboraOnline/online/pull/10512 just "tidies things up" or if this pr on its own isn't sufficient and the other is needed to work properly?

sgothel commented:
> I guess the obvious question then is if #10512 just "tidies things up" or if this pr on its own isn't sufficient and the other is needed to work properly?

Yes. The idle behavior in discussion may even render idle determination useless. In such case we wouldn't care and just sent the invisible tiles at last and one tile per sliced-loop from KitPoll. 
Otherwise, having the idleHint helps to exclude certain idle-determinations when LO is busy, i.e. claiming high-priority.

sgothel commented:
Updated based on timeout (remaining time) only, disregarding potential open-ended idle/wakeup case ([note](https://github.com/CollaboraOnline/online/pull/10512#issuecomment-2487631933))
> Due to ambiguous timeout==0 semantics w/o priority (idle or intentional zero timeout),
> we are not able to determine general idleness w/o passing this information down to COOL.
> (Post reviewing LO's Task, Timer and Scheduler, confirmed by vcl/README.scheduler.md)

Updated description accordingly.

mmeeks commented:
Abandon this for now.
